File: 42. Trapping Rain Water/Trapping Rain Water.java
class Solution {
    public int trap(int[] h) {
        int l = 0;
        int r = h.length - 1;
        int maxL = h[0];
        int maxR = h[r];
        int result = 0;

        while (l < r) {
            if (maxL < maxR) {
                l++;
                maxL = Math.max(maxL, h[l]);
                result += (maxL - h[l]);
            } else {
                r--;
                maxR = Math.max(maxR, h[r]);
                result += (maxR - h[r]);
            }
        }

        return result;
    }
}

File: 42. Trapping Rain Water/Trapping Rain Water.py
class Solution:
    def trap(self, h: List[int]) -> int:
        l = 0
        r = len(h)-1
        maxL = h[0]
        maxR = h[r]
        result = 0
        
        while l < r :
            if maxL < maxR:
                l +=1
                maxL = max(maxL , h[l])
                result += (maxL - h[l])
            else :
                r -=1
                maxR = max(maxR , h[r])
                result += (maxR - h[r])

        return result 

        
            
                

